### ISSUE TITLE
Query parses in postgres but not in redshift

### ISSUE THREAD

**Issue by puetsch** [sqlfluff/sqlfluff#2649]
```sql
select b::datetime
from a

```

This query is properly formatted but result in the message
> Line 1, Position 9: Found unparsable section: '::datetime'

## Expected Behaviour
parses ok

## Observed Behaviour
Error
> Line 1, Position 9: Found unparsable section: '::datetime'

## Steps to Reproduce
`sqlfluff lint --dialect redshift test2.sql`

## Dialect
redshift

## Version
sqlfluff, version 0.10.0
Py 3.8

## Configuration
Note: overwriting the dialect on command line
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = postgres
templater = jinja
rules = None
exclude_rules = L031,L034
recurse = 0
output_line_length = 80
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:indentation]
indented_joins = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 80
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L003]
lint_templated_tokens = True

[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = consistent

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = consistent

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

```

**Comment by PLBMR** [comment]
**TL;DR**: This bug is related to the failure of our Redshift Dialect to recognize `datetime` as a datatype. The short-term solution is to update the Redshift dialect to recognize `datetime` as a datatype. The reason why the Postgres dialect considered `datetime` to be a dataype is because this dialect currently has a very loose definition for datatypes. In the long-term, we should review whether this definition is safe and sustainable for rule management.

# Reproduction

I got the same error as @puetsch with the above `.sqlfluff` file, and the query above written to `test2.sql`:

```bash
mkdir -p ~/.virtualenvs
python3.8 -m venv ~/.virtualenvs/issue-2649-reproduce
source ~/.virtualenvs/issue-2649-reproduce/bin/activate
pip install sqlfluff==0.10.0
...
sqlfluff lint --dialect=redshift test2.sql 
== [test2.sql] FAIL                                                                                                                                           
L:   1 | P:   9 |  PRS | Line 1, Position 9: Found unparsable section:
                       | '::datetime'
All Finished 📜 🎉!
sqlfluff lint --dialect=postgres test2.sql
All Finished 📜 🎉!
```

# Root Cause Investigation

We can see from the parse tree that the Postgres Dialect is able to detect the `datetime` segment as a `data_type_identifier`:

```bash
sqlfluff parse --dialect=postgres test2.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    expression:
[L:  1, P:  8]      |                        column_reference:
[L:  1, P:  8]      |                            identifier:                       'b'
[L:  1, P:  9]      |                        cast_expression:
[L:  1, P:  9]      |                            casting_operator:                 '::'
[L:  1, P: 11]      |                            data_type:
[L:  1, P: 11]      |                                data_type_identifier:         'datetime'
[L:  1, P: 19]      |            newline:                                          '\n'
[L:  2, P:  1]      |            [META] dedent:
[L:  2, P:  1]      |            from_clause:
[L:  2, P:  1]      |                keyword:                                      'from'
[L:  2, P:  5]      |                whitespace:                                   ' '
[L:  2, P:  6]      |                from_expression:
[L:  2, P:  6]      |                    [META] indent:
[L:  2, P:  6]      |                    from_expression_element:
[L:  2, P:  6]      |                        table_expression:
[L:  2, P:  6]      |                            table_reference:
[L:  2, P:  6]      |                                identifier:                   'a'
[L:  2, P:  7]      |                    [META] dedent:
[L:  2, P:  7]      |    newline:                                                  '\n'
```

Whereas the Redshift Dialect does not detect `datetime` as a `data_type_identifier`:

```bash
sqlfluff parse --dialect=redshift test2.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    column_reference:
[L:  1, P:  8]      |                        identifier:                           'b'
[L:  1, P:  9]      |                    unparsable:                               !! Expected: 'Nothing...'
[L:  1, P:  9]      |                        casting_operator:                     '::'
[L:  1, P: 11]      |                        raw:                                  'datetime'
[L:  1, P: 19]      |            newline:                                          '\n'
[L:  2, P:  1]      |            [META] dedent:
[L:  2, P:  1]      |            from_clause:
[L:  2, P:  1]      |                keyword:                                      'from'
[L:  2, P:  5]      |                whitespace:                                   ' '
[L:  2, P:  6]      |                from_expression:
[L:  2, P:  6]      |                    [META] indent:
[L:  2, P:  6]      |                    from_expression_element:
[L:  2, P:  6]      |                        table_expression:
[L:  2, P:  6]      |                            table_reference:
[L:  2, P:  6]      |                                identifier:                   'a'
[L:  2, P:  7]      |                    [META] dedent:
[L:  2, P:  7]      |    newline:                                                  '\n'

==== parsing violations ====
L:   1 | P:   9 |  PRS | Line 1, Position 9: Found unparsable section: '::datetime'
```

The reason for this difference is that Redshift Dialect's [`DatatypeSegment`](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_redshift.py#L90) does not parse `datetime` as a valid `DatatypeSegment`, whereas the Postgres Dialect's `DatatypeSegment` (inherited from the [ANSI Dialect](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_ansi.py#L618)) recognizes `datetime` as a valid [`DatatypeIdentifierSegment`](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_postgres.py#L589) (inherited from the [ANSI Dialect](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_ansi.py#L314)). `DatatypeIdentifierSegment`s are legally `DatatypeSegment`s in the postgres dialect.

The reason why the Redshift Dialect's `DatatypeSegment` does not recognize `datetime` as a valid datatype is likely because we were not aware that `datetime` was a reserved word in Redshift for some time. Redshift did not appropriately document `datetime` as a reserved word in its [spec](https://docs.aws.amazon.com/redshift/latest/dg/r_pg_keywords.html). We addressed this reserved word issue in #2558 .

# Action Items

1. To address this immediate issue, **we will update the Redshift Dialect to properly parse `datetime` as a valid datatype**. Since `datetime` is really a [`DatetimeTypeIdentifier`](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_postgres.py#L452), the way we should make this update is by:
1.1. Updating the Postgres dialect to parse `datetime` as a `DatetimeTypeIdentifier`
1.2. Havre the Redshift dialect inherit this change from the Postgres Dialect
1.3. Update the Redshift `DatatypeSegment` to parse `datetime` as a `DatetimeTypeIdentifier`.

2. In the long-run, we should review the safety of [`DatatypeIdentifierSegment`](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_ansi.py#L314) for parsing datatypes. This segment effectively makes all words that immediately occur after a cast operator (`::`) datatypes. This leads to ANSI and Postgres treating too many words as valid datatypes in their dialects. Because the Redshift dialect did not use the `DatatypeIdentifierSegment` for parsing datatypes, the Redshift dialect's datatype parsing diverged substantially from its inherited dialects (e.g. Postgres, ANSI). I think it's much safer for each dialect to have a stricter range of valid datatypes than the `DatatypeIdentifierSegment` permits. We will address this in a separate issue.

**Comment by PLBMR** [comment]
Small update: [**`datetime` is not a valid type in Postgres**](https://www.postgresql.org/docs/current/datatype-datetime.html). I confirmed this by running the following on an RDS database (up-to-spec with `postgresql`):

```sql
testdb=> SELECT created::datetime FROM example.table LIMIT 1;
ERROR:  type "datetime" does not exist
```

However, according to spec, [**`datetime` is a legal type in `ansi`**](https://www.ibm.com/docs/en/informix-servers/12.10?topic=types-ansi-sql-standards-datetime-interval-values). Winnie Christ, that is confusing.

In this respect, the correct way to solve this bug is by:

1. de-supporting `datetime` as a valid datatype for the Postgres Dialect
2. supporting `datetime` as a valid datatype for the Redshift Dialect.

Will take on this.

**Comment by puetsch** [comment]
How odd! `datetime` indeed is [not documented](https://docs.aws.amazon.com/redshift/latest/dg/r_Datetime_types.html).
In [this part of the postgres documentation](https://www.postgresql.org/docs/6.3/c0804.htm) it sounds like `datetime` might be considered deprecated in lieu of `timestamp`.

I guess in my use case, the answer is probably that I shouldn't use it to start with.

From a product standpoint, the question then becomes whether SQLFluff strives to support language constructs that are considered deprecated by today's standards. Assuming that those constructs still exist in the wild (like in the example I found above), the answer might be "yes".

**Comment by tunetheweb** [comment]
> From a product standpoint, the question then becomes whether SQLFluff strives to support language constructs that are considered deprecated by today's standards. Assuming that those constructs still exist in the wild (like in the example I found above), the answer might be "yes"

I would say we should support deprecated, but not unsupported language constructs. This was dropped in Postgres 7.3 which is over 20 years old now. So I'd be quite comfortable dropping this from our Postgres support (but also adding to Redshift if it is still supported there).

**Comment by tunetheweb** [comment]
> Another alternative would be a validation rule that alerts when deprecated keywords are being used.

Maintaining a list of deprecated words could be troublesome, but you can block words yourself using [rule L062](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.Rule_L062).

**Comment by PLBMR** [comment]
Hey all, a little added complexity: while `postgres` doesn't have a native `DATETIME` datatype in modern versions, you can theoretically construct a [`DATETIME` user-defined data type](https://www.postgresql.org/docs/9.5/xtypes.html), and then tell postgres how to [cast other types that user-defined data type](https://www.postgresql.org/docs/8.4/sql-createcast.html). This is entirely what the `DatatypeIdentifierSegment` is supposed to capture in the Postgres Dialect (indicator [here](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_postgres.py#L588)).

Thus, at the end of the day, the implementation acceptance criteria would be:
1. `DATETIME` continues to be parsed in the Postgres dialect, but as a `DatatypeIdentifierSegment`, instead of as a native datetime type (`DateTimeTypeIdentifier`).
2. `DATETIME` can be parsed properly in the Redshift dialect, but as a native datetime type (`DateTimeTypeIdentifier`).
